### PR TITLE
Core: Clean up state after navigating away

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
Right now most state slices except `dashboard` and `explore` actually clean up their state after navigating away, this can potentially create weird bugs. This PR makes a first attempt to remedy this by being explicit when reducers should clear their state.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/core/redux/resetStateForPath.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Reducer } from 'redux';
+import { ActionOf } from './actionCreatorFactory';
+import { locationChanged, LocationChanged } from '../actions/location';
+
+export const resetStateForPath = <State>(reducer: Reducer<State, ActionOf<any>>, resetStateForPath: string) => (
```
Comment (hugohaggmark) [inline public/app/core/redux/resetStateForPath.ts]: Used for old `switch` type reducers

**Comment by dprokop** [comment]
Interesting idea! I like it. 

But it sort of surfaces my concern about storying transient state in Redux, which requires tricks like this.

File: public/app/features/teams/state/reducers.ts [hunk 2]
```diff
@@ -39,6 +40,6 @@ export const teamReducer = (state = initialTeamState, action: Action): TeamState
 };
 
 export default {
-  teams: teamsReducer,
-  team: teamReducer,
+  teams: resetStateForPath(teamsReducer, '/org/teams'),
+  team: resetStateForPath(teamReducer, '/org/teams'),
```
Comment (torkelo) [inline public/app/features/teams/state/reducers.ts]: Having path here & in router feels fragile, for granted very rare cases when urls change so should be fine. Why is team property being changed here when you move away from teams list?

**Review by torkelo — COMMENTED** [review]
Looks very interesting! Would help in a few cases when we need to clean up state (like transient search filters) 

But there is something that does not feel 100% right about this. Either a smell that we have to do this in the first place. But then again it must a common problem? Are there some common solutions to this? Is this a common solution? 

* Having to specify the path to resetStateForPath (so it's duplicated from router) feels wrong somehow. 
* Maybe easier / clerer to have a generic "rest" action that we can call on mount. But that's too late I think.

**Comment by hugohaggmark** [comment]
> Interesting idea! I like it.
> 
> But it sort of surfaces my concern about storying transient state in Redux, which requires tricks like this.

Good point @dprokop but if I have to choose between having "dumb" components and the truth in state or having "smart" components and nothing in state I would chose the former because it's so much easier to test. But that is my personal preference as you know :)

**Comment by hugohaggmark** [comment]
> Looks very interesting! Would help in a few cases when we need to clean up state (like transient search filters)
> 
> But there is something that does not feel 100% right about this. Either a smell that we have to do this in the first place. But then again it must a common problem? Are there some common solutions to this? Is this a common solution?
> 
Firstly I think the `reducerFactory` is an uncommon solution :) but the need for cleaning up state is not that uncommon I think. We had the same challenge at my last employer.
> * Having to specify the path to resetStateForPath (so it's duplicated from router) feels wrong somehow.
>
Yeah, we could have constants for that.
> * Maybe easier / clerer to have a generic "rest" action that we can call on mount. But that's too late I think.
>
I like that approach a lot, will open another PR based on this.

**Comment by dprokop** [comment]
@hugohaggmark I wonder if there is a chance to put this into the routes definition instead of the particular reducers? So that the action would be dispatched when RectContainer unmounts. Not sure if thats feasible but would keep the logic when it happens.

Route definition would look sth like this:
```ts
.when('/org/teams', {
      template: '<react-container />',
      resolve: {
        resetStateForReducer: teamsReducer, 
        roles: () => (config.editorsCanAdmin ? [] : ['Editor', 'Admin']),
        component: () => SafeDynamicImport(import(/* webpackChunkName: "TeamList" */ 'app/features/teams/TeamList')),
      },
    })
```

**Comment by hugohaggmark** [comment]
> @hugohaggmark I wonder if there is a chance to put this into the routes definition instead of the particular reducers? So that the action would be dispatched when RectContainer unmounts. Not sure if thats feasible but would keep the logic when it happens.
> 
> Route definition would look sth like this:
> 
> ```ts
> .when('/org/teams', {
>       template: '<react-container />',
>       resolve: {
>         resetStateForReducer: teamsReducer, 
>         roles: () => (config.editorsCanAdmin ? [] : ['Editor', 'Admin']),
>         component: () => SafeDynamicImport(import(/* webpackChunkName: "TeamList" */ 'app/features/teams/TeamList')),
>       },
>     })
> ```

That's also a great suggestion, I'm gonna try a different approach in another PR but  I'll come back to this after that.

**Comment by hugohaggmark** [comment]
Closing this in favour of #19392.

**Pull request closed without merging**